### PR TITLE
#26648 Fix "Use a more compact 'Modern' layout" checkbox

### PR DESCRIPTION
**Pull request by rashmitpankhania**
This PR aims to fix `Type: [defect]` [#26648](https://github.com/element-hq/element-web/issues/26648) which is to make "Use a more compact 'Modern' layout" checkbox toggleable.

Understanding:
The value _disabled_ was getting assigned only when the component was mounting the first time that's why on clicking _Show/Hide Advanced_ it was working.
But on switching the layout although it renders the _SettingsFlag_ but the value was not getting updated so instead of taking from state, I decided to take it from SettingsStore.

Further Refactoring??
I think the state _disabled_ may not be needed as its value is sort of getting derived from the flag itself.

## Checklist

-   [ ] Tests written for new code (and old code if feasible)
-   [ ] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * #26648 Fix "Use a more compact 'Modern' layout" checkbox ([\#12090](https://github.com/matrix-org/matrix-react-sdk/pull/12090)). Contributed by @rashmitpankhania.

### CODE REVIEW

File: src/components/views/elements/SettingsFlag.tsx [hunk 1]
```diff
@@ -108,7 +108,7 @@ export default class SettingsFlag extends React.Component<IProps, IState> {
         const label = this.props.label ?? SettingsStore.getDisplayName(this.props.name, this.props.level);
         const description = SettingsStore.getDescription(this.props.name);
         const shouldWarn = SettingsStore.shouldHaveWarning(this.props.name);
-        const disabled = this.state.disabled || !canChange;
+        const disabled = this.isSettingDisabled() || !canChange;
```
Comment (dbkr) [inline src/components/views/elements/SettingsFlag.tsx:111]: Hmm, I'm not convinced by this fix. Isn't the point of the state that it represents the state of the component's UI? If the state is wrong, I think we'd just end up with more problems by working around it.

Comment (rashmitpankhania) [reply src/components/views/elements/SettingsFlag.tsx:111]: I agree your point in general but in this (i guess corner) case specifically

the "disabled" state of checkbox is  established/updated on only two ocassions

1. during mounting of component i.e in constructor
2. when this particular setting gets changed... but here is the catch "useCompactLayout" settings uses an   Incompatible controller, `controller: new IncompatibleController("layout", false, (v: Layout) => v !== Layout.Group)`, and disabled value depends on "layout" settings

so anytime the "layout" settings changes the value of disabled changes but state of the checkbox doesnt know about it...thats why i put the to get correct value on every render i replaced the state with the function call.

I think the "canChange" in line 104 SettingsFlag.tsx `const canChange = SettingsStore.canSetValue(this.props.name, this.props.roomId ?? null, this.props.level);` is somewhat analogous to "disabled" where its getting value directly from SettingsStore not as state.

So in a nutshell, may be we need to handle it special or any other approach from you is also welcome. 
Thanks for the response.
